### PR TITLE
Serialize functions with kryo

### PR DESCRIPTION
**Pull request by kul**
I have not seen this mentioned anywhere, but we have an if condition in `gen-function` macro which will let you pass plain function in transformations as well. So you can pass aot compiled functions directly without any problems e.g.
`(f/map inc)` and `(f/map (fn [x] (+ x x))` is completely valid if aot compiled.

This pr inhances this to serialize plain functions with kryo instead. Which has added advantage that it lets you close over locals too. It requires aot so cannot be used directly over a repl which has sc over a remote spark cluster but works fine on local spark cluster.

### CODE REVIEW

**Comment by kul** [comment]
This pr does not break/change anything. But it lets you use plain clojure functions for local spark dev and production without overhead of `serializable-fn`. For dev over a remote spark cluster you can still use `f/fn`.

**Comment by sorenmacbeth** [comment]
What sort of overhead are you experiencing using serializable-fn locally? We use it here and really haven't noticed anything. 

I'm just not sure the benefit is worth the potential confusion around when it's ok to use one or the other. 

Could you elaborate a bit more on why you don't want to use the serializable-fns locally?

File: src/clojure/flambo/function.clj [hunk 1]
```diff
@@ -52,8 +52,10 @@
         :constructors {[Object] []})
        (defn ~wrapper-name [f#]
          (new ~new-class-sym
-              (if (serfn? f#) (binding [sfn/*serialize* kryo/serialize]
-                                (serialize-fn f#)) f#))))))
+              (if (serfn? f#)
+                (binding [sfn/*serialize* kryo/serialize]
+                  (serialize-fn f#))
+                [(kryo/serialize f#)]))))))
```
Comment (sorenmacbeth) [inline src/clojure/flambo/function.clj]: why are you wrapping the function in a vector here?

**Comment by kul** [comment]
@sorenmacbeth To be honest i have not done any performance tests yet neither am I advocating use of plain functions if the api user do not desire so. This PR just enhances what is already there.

And this would clear `serializable-fn` of the accusations of causing perm-gen at least :laughing:

File: src/clojure/flambo/function.clj (hunk 1 above)
Comment (kul) [inline src/clojure/flambo/function.clj]: This is just to make a distinction between serializable fn and plain fns in `-call`'s if condition
`(instance? array-of-bytes-type fn-or-serfn)`.

**Comment by kul** [comment]
I would be squashing in one more commit to memoize deserialization of plain functions.

**Comment by kul** [comment]
Squashed in a new commit. Let me know if this makes sense.

**Comment by kul** [comment]
Been using this in production for a while without any problems. Any words on this?

File: src/clojure/flambo/function.clj [hunk 2]
```diff
@@ -11,7 +11,11 @@
   (= (type f) :serializable.fn/serializable-fn))
 
 (def serialize-fn sfn/serialize)
-(def deserialize-fn (memoize sfn/deserialize))
+(def deserialize-serfn (memoize sfn/deserialize))
+(def deserialize-fn (memoize #(let [f (kryo/deserialize (first %))]
+                                (->> f class .getName (re-matches #"(.*?)\$.*")
```
Comment (sorenmacbeth) [inline src/clojure/flambo/function.clj]: I'm concerned about having a regular expression here. I realize the function is memoize, but it's still concerning performance-wise.

Comment (kul) [inline src/clojure/flambo/function.clj]: yea.. i dont like that as well. That is basically to get the namespace for an unbound function and require it.

**Comment by kul** [comment]
Remove the regex for ns resolution. WDYT?

**Pull request closed without merging**